frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2618,24 +2745,200 @@ lookupCalledExprConsideringReceiver(Context* context,
   return ret;
 }
 
+static void helpComputeForwardingTo(const CallInfo& fci,
+                                    size_t start,
+                                    CandidatesVec& candidates,
+                                    std::vector<QualifiedType>& forwardingTo) {
+  QualifiedType forwardingReceiverActualType = fci.calledType();
+  size_t n = candidates.size();
+  forwardingTo.resize(start);
+  for (size_t i = start; i < n; i++) {
+    forwardingTo.push_back(forwardingReceiverActualType);
+  }
+}
+
+// this function gathers candidates not from POI and candidates
+// from POI into separate vectors.
+// For each of these vectors, the corresponding forwardingTo vector
+// will have an element for each of the returned candidates &
+// indicates the actual type that is passed as the method receiver
+// when using forwarding.
+static void
+gatherAndFilterCandidatesForwarding(Context* context,
+                                    const Call* call,
+                                    const CallInfo& ci,
+                                    const Scope* inScope,
+                                    const PoiScope* inPoiScope,
+                                    CandidatesVec& nonPoiCandidates,
+                                    CandidatesVec& poiCandidates,
+                                    ForwardingInfoVec& nonPoiForwardingTo,
+                                    ForwardingInfoVec& poiForwardingTo) {
+  nonPoiCandidates.empty();
+  poiCandidates.empty();
+  nonPoiForwardingTo.empty();
+  poiForwardingTo.empty();
+
+  const Type* receiverType = ci.actual(0).type().type();
+
+  // Resolve the forwarding expression's type
+  const ResolvedForwarding* forwards = nullptr;
+  UniqueString name = ci.name();
+  if (name == USTR("init") || name == USTR("init=") || name == USTR("deinit")) {
+    // these are exempt from forwarding
+  } else if (auto ct = receiverType->getCompositeType()) {
+    forwards = forwardingForTypeDecl(context, ct);
+  }
+
+  if (forwards) {
+    // Construct CallInfos with the receiver replaced for each
+    // of the forwarded-to types.
+    std::vector<CallInfo> forwardingCis;
+
+    int numForwards = forwards->numForwards();
+    for (int i = 0; i < numForwards; i++) {
+      QualifiedType forwardType = forwards->receiverType(i);
+      std::vector<CallInfoActual> actuals;
+      // compute the actuals
+      // first, the method receiver (from the forwarded type)
+      actuals.push_back(CallInfoActual(forwardType, USTR("this")));
+      // then, the other actuals, if any, from ci
+      int numActuals = ci.numActuals();
+      for (int j = 1; j < numActuals; j++) {
```

Comment:
Maybe a silly suggestion, but I'd ASSERT here that ci.actual(0) is named `this`.